### PR TITLE
Replaced deprecated function

### PR DESCRIPTION
**Pull request by beaubbe**
The function deleteByQuery has been deprecated. https://github.com/algolia/algoliasearch-client-javascript/wiki/Deprecated#indexdeletebyquery
We must now use deleteBy. After running the crawler with the new function, everything seems to be working fine

### CODE REVIEW

**Comment by rayrutjes** [comment]
Hey @beaubbe,
I'd like to share a bit of context here.

`deleteByQuery` has been deprecated because the task previously relied on a client side `browser` of the index.

With `deleteBy` however, the task will be queued in Algolia so that you do not have to worry about it.

Please note however that the signature of the new method is not the same: https://github.com/algolia/algoliasearch-client-javascript/blob/6a318c9f4ae7645bd727f493cb69bae894ea01f6/src/Index.js#L380

I don't think your current implementation will do what you would expect.

I also do realize that our documentation is a bit incomplete and even our JSdocs.

I will make sure we address it.

**Comment by beaubbe** [comment]
Thank you. I will check it

**Comment by beaubbe** [comment]
hey @rayrutjes,
I tried correcting the call to the method by moving the the query '*' in the params object like so:
`{
				query: '*',
				numericFilters: ['timestamp<' + (new Date().getTime() - config.oldentries)]
			}`

Now, i'm having the following error: AlgoliaSearchError: Method not allowed with this API key.
However, I have confirmed that the api keu used can delete records. 
Any idea why I'm having this problem?

**Comment by rayrutjes** [comment]
The query should be removed all together because it is no longer a valid parameter.
I updated the docblock a few minutes ago to reflect this: https://github.com/algolia/algoliasearch-client-javascript/pull/609#event-1290929202

Let me know if that solves your issue.

**Comment by rayrutjes** [comment]
So you should do: 
```js
{ numericFilters: ['timestamp<' + (new Date().getTime() - config.oldentries)] }
```

**Comment by beaubbe** [comment]
Everything seems to be working fine now. Thank you.

However, it seems that my api key now needs to be able to delete indexes to be able to delete my records. 

Before switching to deleteBy, the api key only needed to have "delete records" to be able to delete the records. 

Now, with deleteBy, the api key needs to have "delete index" to be able to delete records. Without "delete index", I get the following error: Method not allowed with this API key.

Is it supposed to be that way?

**Comment by rayrutjes** [comment]
Hey @beaubbe ,

I'm glad it works now.
I will ask about the permission scope, I would expect the same as you.

I'll keep you posted.

**Comment by rayrutjes** [comment]
Hey again @beaubbe .
The fact that you now need the `deleteIndex` permission scope is an intended move.
It reflects the fact that you could eventually loose all information in your index.

I hope that makes sense!

**Comment by nitriques** [comment]
Hi @rayrutjes ! Thanks for shimming in so fast ;) This is nice.

> I hope that makes sense!

It does not. I also think it can break some existing code.

Let's say the user now actually DROPs the index. She'll get 'index not found errors'. Not giving the drop permission mean that when setting index property, even if the user did delete all records, it would still work. Index could not be deleted, never.

1. What does DELETE allows then ? Delete single objects ?
1. Maybe this is what we need here: it's a purge based on the timestamp ?
1. Do I need to update my code to treat the non existing index everywhere ?

Thanks. Your service rocks.

**Comment by rayrutjes** [comment]
Hi @nitriques ,
> I also think it can break some existing code.

Do you have any example in mind?
Please note that we do not replace an existing method. `deleteBy` is a new method that replaces `deleteByQuery`.

> Not giving the drop permission mean that when setting index property, even if the user did delete all records, it would still work. Index could not be deleted, never.

I understand your point of view. However, from a business perspective, an empty index is a deleted index. That is the reason when strengthened the permission scope.

> 1. What does DELETE allows then ? Delete single records ?

Yes, or batch of records.

> 2. Maybe this is what we need here: it's a purge based on the timestamp ?

Yes, but you will not be able to leverage `deleteBy` if not having the `delete index` permission.

> 3. Do I need to update my code to treat the non existing index everywhere ?

If you want to code very defensively, yes? But in the end, your key shouldn't leak so you are the only one being able to delete the index right? And it would be a separate call, because `deleteBy` can only swipe everything.


`deleteBy` is a powerful way of deleting bunch of items. In a single API call you are now able to basically swipe the entire index, regardless of its size.
Previously, there was a loop over the records to delete them by batches.
New implementation, new permission scope.

*with great power comes great responsibility*

**Comment by nitriques** [comment]
> Please note that we do not replace an existing method. deleteBy is a new method that replaces deleteByQuery.

This is true as long as `deleteByQuery` still works. When it does not, code will break.

> Do you have any example in mind?

Yes, all of the existing cron job with the current project configured. I have lots of them, and some are running some pretty old versions.

> Yes, or batch of records.

Ok so I should issue the query, and then delete them based on their IDs ? This would not require the Drop index ? BTW, is this is possible, I could query the ENTIRE index and delete them one by one ?

**Comment by rayrutjes** [comment]
Hi @nitriques ,

If you have multiple instances of your project and some of them are running old versions of the JS client, then you should not upgrade to the new method.
However, if you are to use the new one you will indeed need to upgrade the client everywhere.

Regarding the deprecated method, we will only definitely remove it in a major release, as long as you lock your versions you shouldn't have any issue.
Regarding the Algolia API itself, the old `deleteByQuery`does not rely on any deprecated API endpoint. Which means that if you keep the old implementation and lock your client version, it will never stop to work.

Let me know if you have any question.

tl;dr:  if you don't see any problem with the current implementation, you can keep it. If you realize that sometimes some items that should be deleted aren't then you should migrate to the new `deleteBy` method.

**Comment by nitriques** [comment]
> as long as you lock your versions you shouldn't have any issue.

That's relief.

But for the future, I only see three options, and I do not like one of them:

1. Allow the key to drop the index, which feels less secure to me. ALSO, allow the key to list indexes to check the existence of the index. +1 op per run.
1. Delete objects one by one. +100000 ops per run. Nice for your revenues, not for my clients (which are also your clients!).

I feel like the only solution for this project (and all it running instances) is #1. 

I fully understand your concerns and reverting to the old behaviour is not an option. But please consider creating a new role, i.e. can delete object in batch, but not the index itself. This would at least make it impossible for the index to be dropped and client side code won't throw an error instead of simply returning no results, in the avent someone accidentally drops the index.

Thanks for your help. It's a pleasure to be able to discuss this issue with you. I'll be adding a new feature in the next days, so expect a new release soon ;)

**Comment by nitriques** [comment]
@beaubbe can you add something in the README about the required roles for this project to run successfully please ? Thanks.
